draft-ietf-tls-esni.md

```diff
@@ -989,6 +960,80 @@ The "ech_is_inner" is expected to have an empty payload. If the payload is
 non-empty (i.e., the length of the "extension_data" field is non-zero) then the
 backend server MUST abort the handshake with an "illegal_parameter" alert.
 
+## Handling HelloRetryRequest {#server-hrr}
+
+The HelloRetryRequest constructed by the backend server is called the
+HelloRetryRequestInner message; and the HelloRetryRequest constructed by the
+client-facing server is called the HelloRetryRequestOuter message.
+
+If the backend server cannot complete the handshake with the parameters provided
+in the ClientHelloInner, then it constructs a HelloRetryRequestInner message in
+the usual way and forwards it to the client-facing server. The client-facing
+server encrypts this message using a key derived from the HPKE context, as
+described below. It then constructs HelloRetryRequestOuter message with an
+"encrypted_client_hello" extension containing the ciphertext as its payload. It
+then continues the handshake by sending this message to the client.
+
+The payload of the "encrypted_client_hello" extension is constructed as follows.
+The client-facing server begins by exporting a key and nonce for AEAD
+encryption:
+
+~~~
+    hrr_inner_key = context.Export("hrr inner key", Nk)
+    hrr_inner_nonce = context.Export("hrr inner nonce", Nn)
+~~~
+
+Here `Nk` and `Nn` denote, respectively, the key and nonce length of the AEAD
+algorithm associated with `context`. The extension payload is computed as
+
+~~~
+    encrypted_hrr_inner = Seal(hrr_inner_key,
+                               hrr_inner_nonce,
+                               "", // Empty AAD
+                               HelloRetryRequestInner)
+~~~
+
+The ciphertext `encrypted_hrr_inner` is used as the payload of the
+"encrypted_client_hello" extension. When sending this extension, the
+HelloRetryRequestOuter MAY also contain a "cookie" extension, but SHOULD NOT
+contain any other extension.
```

Comment:
```suggestion
contain any other extension. If present, the contents of the "cookie" extension MUST 
NOT contain any information included in or derived from ClientHelloInner.
```

Reply:
Why is this needed? How would it be enforced?

Reply:
I don't think it can be enforced, but you can imagine a server very stupidly putting the InnerCH in the cookie and sending it back to the client (in plaintext!)

Reply:
Ack. No change made, but we can discuss it in the actual PR.